From 6d63bab7808db0ddb7623dd90935f36f76f5acf6 Mon Sep 17 00:00:00 2001
From: Henry Schreiner <henry.fredrick.schreiner@cern.ch>
Date: Fri, 18 Dec 2020 12:35:49 -0500
Subject: [PATCH] refactor: modernize setup (#231)

* refactor: modernize setup

* tests: pip install

* fix: typo on supported Pythons

* Apply suggestions from code review

* fix: drop other single-line open
---
 .github/workflows/build-test.yml |  3 ++
 pyproject.toml                   |  7 ++++
 setup.cfg                        | 49 ++++++++++++++++++++++++++
 setup.py                         | 60 ++++----------------------------
 4 files changed, 65 insertions(+), 54 deletions(-)
 create mode 100644 pyproject.toml

diff --git a/pyproject.toml b/pyproject.toml
new file mode 100644
index 0000000..e26da06
--- /dev/null
+++ b/pyproject.toml
@@ -0,0 +1,7 @@
+[build-system]
+requires = [
+    "setuptools>=42",
+    "wheel",
+]
+
+build-backend = "setuptools.build_meta"
diff --git a/setup.cfg b/setup.cfg
index f6f2e57..9840f28 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -1,3 +1,52 @@
+[metadata]
+name = uproot
+author = Jim Pivarski
+author_email = pivarski@princeton.edu
+maintainer = Jim Pivarski
+maintainer_email = pivarski@princeton.edu
+description = ROOT I/O in pure Python and NumPy.
+long_description = file: README.md
+long_description_content_type = text/markdown
+url = https://github.com/scikit-hep/uproot4
+download_url = https://github.com/scikit-hep/uproot4/releases
+license = BSD 3-clause
+
+classifiers =
+    Development Status :: 5 - Production/Stable
+    Intended Audience :: Developers
+    Intended Audience :: Information Technology
+    Intended Audience :: Science/Research
+    License :: OSI Approved :: BSD License
+    Operating System :: MacOS
+    Operating System :: POSIX
+    Operating System :: Unix
+    Programming Language :: Python
+    Programming Language :: Python :: 2.6
+    Programming Language :: Python :: 2.7
+    Programming Language :: Python :: 3.5
+    Programming Language :: Python :: 3.6
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.8
+    Programming Language :: Python :: 3.9
+    Topic :: Scientific/Engineering
+    Topic :: Scientific/Engineering :: Information Analysis
+    Topic :: Scientific/Engineering :: Mathematics
+    Topic :: Scientific/Engineering :: Physics
+    Topic :: Software Development
+    Topic :: Utilities
+
+
+[options]
+python_requires = >=2.6, !=3.0, !=3.1, !=3.2, !=3.3, !=3.4
+packages = find:
+install_requires =
+    numpy
+
+[options.packages.find]
+exclude =
+    tests
+
+
 [flake8]
 ignore = E203, W503, E501, E266, N
 max-complexity = 100
diff --git a/setup.py b/setup.py
index 5c03a0b..8140cac 100644
--- a/setup.py
+++ b/setup.py
@@ -2,12 +2,12 @@
 
 import os.path
 from setuptools import setup
-from setuptools import find_packages
 
 
 def get_version():
     g = {}
-    exec(open(os.path.join("uproot", "version.py")).read(), g)
+    with open(os.path.join("uproot", "version.py")) as f:
+        exec(f.read(), g)
     return g["__version__"]
 
 
@@ -17,54 +17,7 @@ def get_version():
 }
 extras["all"] = sum(extras.values(), [])
 
-install_requires = open("requirements.txt").read().strip().split("\n")
-
-setup(name = "uproot",
-      packages = find_packages(exclude = ["tests"]),
-      scripts = [],
-      version = get_version(),
-      author = "Jim Pivarski",
-      author_email = "pivarski@princeton.edu",
-      maintainer = "Jim Pivarski",
-      maintainer_email = "pivarski@princeton.edu",
-      description = "ROOT I/O in pure Python and NumPy.",
-      long_description = open("README.md").read(),
-      long_description_content_type = "text/markdown",
-      url = "https://github.com/scikit-hep/uproot4",
-      download_url = "https://github.com/scikit-hep/uproot4/releases",
-      license = "BSD 3-clause",
-      test_suite = "tests",
-      python_requires = ">=2.6, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*",
-      install_requires = ["numpy"],
-      tests_require = extras["test"],
-      extras_require = extras,
-
-      classifiers = [
-#         "Development Status :: 1 - Planning",
-#         "Development Status :: 2 - Pre-Alpha",
-#         "Development Status :: 3 - Alpha",
-#         "Development Status :: 4 - Beta",
-          "Development Status :: 5 - Production/Stable",
-#         "Development Status :: 6 - Mature",
-#         "Development Status :: 7 - Inactive",
-          "Intended Audience :: Developers",
-          "Intended Audience :: Information Technology",
-          "Intended Audience :: Science/Research",
-          "License :: OSI Approved :: BSD License",
-          "Operating System :: MacOS",
-          "Operating System :: POSIX",
-          "Operating System :: Unix",
-          "Programming Language :: Python",
-          "Programming Language :: Python :: 2.6",
-          "Programming Language :: Python :: 2.7",
-          "Programming Language :: Python :: 3.5",
-          "Programming Language :: Python :: 3.6",
-          "Programming Language :: Python :: 3.7",
-          "Programming Language :: Python :: 3.8",
-          "Topic :: Scientific/Engineering",
-          "Topic :: Scientific/Engineering :: Information Analysis",
-          "Topic :: Scientific/Engineering :: Mathematics",
-          "Topic :: Scientific/Engineering :: Physics",
-          "Topic :: Software Development",
-          "Topic :: Utilities",
-          ])
+setup(
+    version = get_version(),
+    extras_require = extras,
+)
